public/icvar.h

```diff
@@ -11,21 +11,254 @@
 #endif
 
 #include "appframework/IAppSystem.h"
-#include "tier1/convar.h"
 #include "tier1/utlvector.h"
 #include "tier0/memalloc.h"
+#include "mathlib/vector4d.h"
+#include <cstdint>
 
-class ConCommandBase;
+class BaseConVar {};
 class ConCommand;
-class ConVar;
-class Color;
+class CCommand;
+class CCommandContext;
+class ICVarListenerCallbacks;
 class IConVarListener;
-class CConVarDetail;
+struct ConVarCreation_t;
+struct ConCommandCreation_t;
 struct ConVarSnapshot_t;
-union CVValue_t;
 class KeyValues;
-class ConVarRefAbstract;
 
+template<typename T>
+class ConVar;
+
+union CVValue_t
+{
+	CVValue_t() { memset(this, 0, sizeof(*this)); }
+	CVValue_t(CVValue_t const& cp) { memcpy(this, &cp, sizeof(*this)); };
+	CVValue_t& operator=(CVValue_t other) { memcpy(this, &other, sizeof(*this)); return *this; }
+
+	template<typename T>
+	inline CVValue_t& operator=(T other);
+
+	inline operator bool() const			{ return m_bValue; }
+	inline operator int16_t() const			{ return m_i16Value; }
+	inline operator uint16_t() const		{ return m_u16Value; }
+	inline operator int32_t() const			{ return m_i32Value; }
+	inline operator uint32_t() const		{ return m_u32Value; }
+	inline operator int64_t() const			{ return m_i64Value; }
+	inline operator uint64_t() const		{ return m_u64Value; }
+	inline operator float() const			{ return m_flValue; }
+	inline operator double() const			{ return m_dbValue; }
+	inline operator const char*() const		{ return m_szValue; }
+	inline operator const Color&() const	{ return m_clrValue; }
+	inline operator const Vector2D&() const	{ return m_vec2Value; }
+	inline operator const Vector&() const	{ return m_vec3Value; }
+	inline operator const Vector4D&() const	{ return m_vec4Value; }
+	inline operator const QAngle&() const 	{ return m_angValue; }
+
+	bool		m_bValue;
+	int16_t		m_i16Value;
+	uint16_t	m_u16Value;
+	int32_t		m_i32Value;
+	uint32_t	m_u32Value;
+	int64_t		m_i64Value;
+	uint64_t	m_u64Value;
+	float		m_flValue;
+	double		m_dbValue;
+	const char*	m_szValue;
+	Color		m_clrValue;
+	Vector2D	m_vec2Value;
+	Vector		m_vec3Value;
+	Vector4D	m_vec4Value;
+	QAngle		m_angValue;
+};
+static_assert(sizeof(float) == sizeof(int32_t), "Wrong float type size for ConVar!");
+static_assert(sizeof(double) == sizeof(int64_t), "Wrong double type size for ConVar!");

```

Comment:
I'd like for these to stay, since `float` and `double` are not thoroughly defined by the C++ standard for the amount of bytes they occupy. And its very that their size does not change when it comes to EConVarType_Float32/64 respectively.